Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx

```diff
@@ -314,6 +326,176 @@ bool vtkCurveMeasurementsCalculator::CalculatePolyDataCurvature(vtkPolyData* pol
+  if (!curveCoordinateSystemGenerator)
+    {
+    vtkErrorMacro("Failed to access coordinate system generator");
+    return false;
+    }
+  vtkPointData* pointData = polyData->GetPointData();
+  vtkDoubleArray* tangents = vtkDoubleArray::SafeDownCast(
+    pointData->GetAbstractArray(curveCoordinateSystemGenerator->GetTangentsArrayName()));
+  vtkDoubleArray* binormals = vtkDoubleArray::SafeDownCast(
+    pointData->GetAbstractArray(curveCoordinateSystemGenerator->GetBinormalsArrayName()));
+  // Sanity checks
+  vtkPolyData* curvePoly = curveCoordinateSystemGenerator->GetOutput();
+  if (curvePoly->GetNumberOfPoints() != polyData->GetNumberOfPoints())
+    {
+    // Sanity check
+    vtkErrorMacro("Number of points does not match between given poly data and the coordinate system generator poly data ("
+      << curvePoly->GetNumberOfPoints() << " != " <<  polyData->GetNumberOfPoints() << ")");
+    return false;
+    }
+  if (binormals->GetNumberOfTuples() != pointData->GetNumberOfTuples())
+    {
+    vtkErrorMacro("Number of data points does not match between coordinate system generator ("
+      << binormals->GetNumberOfTuples() << ") and poly data point data (" << pointData->GetNumberOfTuples() << ")");
+    return false;
+    }
+
+  vtkCellArray* lines = polyData->GetLines();
+  vtkNew<vtkIdList> linePoints;
+
+  lines->GetCell(0, linePoints);
+  vtkIdType numberOfPoints = // Last point in closed curve line is the first point
+    (this->CurveIsClosed ? linePoints->GetNumberOfIds()-1 : linePoints->GetNumberOfIds());
+
+  // Initialize torsion array
+  vtkNew<vtkDoubleArray> torsionArray;
+  torsionArray->Initialize();
+  torsionArray->SetName("Torsion");
+  torsionArray->SetNumberOfComponents(1);
+  torsionArray->SetNumberOfTuples(numberOfPoints);
+  torsionArray->Reset();
+  torsionArray->FillComponent(0,0.0);
+
+  // Initialize curvature variables
+  double minTorsion= 0.0;
+  double maxTorsion = 0.0;
+  double meanTorsion = 0.0;
+
+  // Get values for first point
+  double* binormal = binormals->GetTuple3(linePoints->GetId(1));
+  double binormalNorm = sqrt(binormal[0]*binormal[0] + binormal[1]*binormal[1] + binormal[2]*binormal[2]);
+  double normBinormal[3] = {0.0, 0.0, 0.0};
+  double prevNormBinormal[3] = {binormal[0]/binormalNorm, binormal[1]/binormalNorm, binormal[2]/binormalNorm};
+  double torsion = 0.0;
+  double *tangent = nullptr;
+  double currentLength = 0.0;
+  double length = 0.0;
+
+  // The torsion for the first cell is 0.0 for open curves
+  torsionArray->InsertValue(linePoints->GetId(0), 0.0);
+
+  for (vtkIdType idx=1; idx<numberOfPoints-1; ++idx)
+    {
+    binormal = binormals->GetTuple3(linePoints->GetId(idx+1));
+    binormalNorm = sqrt(binormal[0]*binormal[0] + binormal[1]*binormal[1] + binormal[2]*binormal[2]);
+
+    normBinormal[0] = binormal[0] / binormalNorm;
+    normBinormal[1] = binormal[1] / binormalNorm;
+    normBinormal[2] = binormal[2] / binormalNorm;
+
+    // Local torsion
```

Comment:
could you add a reference to where the computation comes from?

Reply:
Done